File: CgmLogger/src/main/java/com/kludgenics/cgmlogger/data/logdata/location/api/GooglePlacesWebLocation.java
package com.kludgenics.cgmlogger.data.logdata.location.api;

import com.google.gson.annotations.Expose;
import com.kludgenics.cgmlogger.data.logdata.location.data.GeocodedLocation;
import com.kludgenics.cgmlogger.data.logdata.location.data.Position;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

/**
 * Created by matthiasgranberry on 5/23/15.
 */
public class GooglePlacesWebLocation implements GeocodedLocation {
    @Expose
    String place_id;
    @Expose
    List<String> types;
    @Expose
    String name;
    @Expose
    Geometry geometry;

    class Geometry {
        @Expose
        GeometryLocation location;
        Position position;

        Position getPosition() {
            if (position == null) {
                position = new Position(location.lat, location.lng);
            }
            return position;
        }

        class GeometryLocation {
            @Expose
            float lat;
            @Expose
            float lng;
        }
    }
    @Override
    public String getId() {
        return place_id;
    }

    @Override
    public CharSequence getName() {
        return name;
    }

    @Override
    public String getLocationTypes() {
        return null;
    }

    @Override
    public CharSequence getAddress() {
        return StringUtils.join(types, '|');
    }

    @Override
    public Position getPosition() {
        return geometry.getPosition();
    }

    @Override
    public CharSequence getAttributionSnippet() {
        return null;
    }
}

File: app/src/main/java/com/kludgenics/cgmlogger/model/location/places/GooglePlacesWebLocation.kt
package com.kludgenics.cgmlogger.model.location.places

import com.google.gson.annotations.Expose
import com.kludgenics.cgmlogger.model.location.data.GeocodedLocation
import com.kludgenics.cgmlogger.model.location.Position
import java.util.Collections

/**
 * Created by matthiasgranberry on 5/23/15.
 */
public class GooglePlacesWebLocation : GeocodedLocation {
    @Expose
    var place_id: String = ""
    @Expose
    var types: List<String> = Collections.emptyList()
    @Expose
    var name: String = ""
    @Expose
    var geometry: Geometry = Geometry()

    inner class Geometry {
        @Expose
        var location: GeometryLocation = GeometryLocation()

        fun getPosition(): Position {
            return Position(location.lat.toDouble(), location.lng.toDouble())
        }

        inner class GeometryLocation {
            @Expose
            var lat: Float = 0.toFloat()
            @Expose
            var lng: Float = 0.toFloat()
        }
    }

    override fun getId(): String {
        return place_id
    }

    override fun getName(): CharSequence {
        return name
    }

    override fun getLocationTypes(): String {
        return types.joinToString("|")
    }

    override fun getAddress(): CharSequence {
        return ""
    }

    override fun getPosition(): Position {
        return geometry.getPosition()
    }

    override fun getAttributionSnippet(): CharSequence? {
        return null
    }
}